delta_backend/src/extractor.py

```diff
@@ -0,0 +1,409 @@
+import json
+import exception_messages
+from datetime import datetime, timedelta, timezone
+from common.mappings import Gender
+
+class Extractor: 
+
+    # This file holds the schema/base layout that maps FHIR fields to flat JSON fields
+    # Each entry tells the converter how to extract and transform a specific value
+    EXTENSION_URL_VACCINATION_PRODEDURE = "https://fhir.hl7.org.uk/StructureDefinition/Extension-UKCore-VaccinationProcedure"
+    EXTENSION_URL_SCT_DESC_DISPLAY = "https://fhir.hl7.org.uk/StructureDefinition/Extension-UKCore-CodingSCTDescDisplay"
+
+    CODING_SYSTEM_URL_SNOMED = "http://snomed.info/sct"
+    ODS_ORG_CODE_SYSTEM_URL = "https://fhir.nhs.uk/Id/ods-organization-code"
+    DEFAULT_LOCATION = "X99999"
+    NHS_NUMBER_SYSTEM_URL = "https://fhir.nhs.uk/Id/nhs-number"
+    
+    DATE_CONVERT_FORMAT = "%Y%m%d"
+    DEFAULT_POSTCODE = "ZZ99 3CZ"
+    
+    def __init__(self, fhir_json_data, report_unexpected_exception = True):
+        self.fhir_json_data = json.loads(fhir_json_data) if isinstance(fhir_json_data, str) else fhir_json_data
+        self.report_unexpected_exception = report_unexpected_exception
+        self.error_records = []
+        
+    def _get_patient(self):
+        contained = self.fhir_json_data.get("contained", [])
+        return next((c for c in contained if isinstance(c, dict) and c.get("resourceType") == "Patient"), "")
+
+    def _get_valid_names(self, names, occurrence_time):  
+        official_names = [n for n in names if n.get("use") == "official" and self._is_current_period(n, occurrence_time)]
+        if official_names:
+            return official_names[0]
+
+        valid_names = [n for n in names if self._is_current_period(n, occurrence_time) and n.get("use") != "old"]
+        return valid_names[0] if valid_names else names[0]
+    
+    def _get_person_names(self):
+        occurrence_time = self._get_occurance_date_time()
+        patient = self._get_patient()
+        names = patient.get("name", [])
+        
+        if not isinstance(names, list) or not names:
+            return "", ""
+
+        selected_name = self._get_valid_names(names, occurrence_time)
+        person_forename = " ".join(selected_name.get("given", []))
+        person_surname = selected_name.get("family", "")
+
+        return person_forename, person_surname
+    
+    def _get_practitioner_names(self):
+        contained = self.fhir_json_data.get("contained", [])
+        occurrence_time = self._get_occurance_date_time()
+        practitioner = next((c for c in contained if isinstance(c, dict) and c.get("resourceType") == "Practitioner"), None)
+        if not practitioner or "name" not in practitioner:
+            return "", ""
+
+        practitioner_names = practitioner.get("name", [])
+        valid_practitioner_names = [n for n in practitioner_names if "given" in n or "family" in n]
```

Comment:
The spec says we should have a `given` AND `family`. I feel like we should question this though - follow up ticket?